### DISCUSSION TITLE
I am facing performance issues with Prisma with MONGODB. How can I improve the performance?

### DISCUSSION THREAD

**Discussion by AkashVarmaGadiraju** [prisma/prisma#22146] [Q&A]
### Question

I used to use only MONGODB as database and mongoose in NodeJs to access it. I recently started using POSTGRES and MONGODB both in my recent project and then learnt about wonders of PRISMA. I integrated it into my project and created schemas. Once I started using it I noticed that MONGODB crud operations were exponentially slower than mongoose calls.

Few of the reasons I think might the issue be are:

   1.  For Prisma to support NoSQL databases, they added an extra layer of validation, this extra layer is slowing it down.

   2. Every time I run a query will it create a new connection with database and then perform action. If so that is an expensive operation.

   3. We might have many indexes or uniques which might slow down the database.

These are few of the reasons that I can think.

I tried creating a 1000 documents using PRISMA and Mongoose. The results are:

> PRISMA took 1 min 31 sec
> Mongoose took 10.91 sec

I tried creating 1000 records in POSTGRES. The result is:

> PRISMA took 4.37 sec

I expected the output to be faster in Prisma with MONGODB because I will be saving something like transactions which happen quiet often in this environment. Hence if there is a lag then it will slowdown the whole server.

This is the dummy code which I used to perform testing using prisma:

```ts
let transactions: any = [];
for (let i = 0; i < bulk.quantity; ++i)
{
	transactions.push(
		await mongoClient.transaction.create({
			data: {
				topic: bulk.topicName + i,
				messagingPolicyId: bulk.messagingPolicyId,
			},
		})
	);
}
let response = await Promise.all(transactions);
serverLogger.info(transactions.length + " Transactions Successfully Generated");
return response;
```

I used thunder client to benchmark these results and performed API calls to test it.

What might me the possible resolutions for this problem?

This is my first time posting in GitHub, so if I missed any detail or gave you a bad explanation, please do guide me. Thanks fro helping.

### Expected behavior (optional)

The performance difference between mongoose and prism should reduce.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma

model Transaction {
  id                String          @id @default(auto()) @map("_id") @db.ObjectId
  topic             String
  QOS               Int?            @default(0) //TODO: Needs to be Review
  payload           Json?
  messagingPolicyId String          @db.ObjectId
  actionId          String?         @db.ObjectId
  metadata          Json?
  MessagingPolicy   MessagingPolicy @relation(fields: [messagingPolicyId], references: [id], onDelete: Cascade)
  createdAt         DateTime        @default(now())
  updatedAt         DateTime        @updatedAt
}

// Add your schema.prisma
```

- OS: Tested on Windows and macOs
- Database: MongoDB
- Prisma: 5.4.2
- TypeScript: 5
- @types/express: ^4.17.19,
- @types/node: ^20.8.4

**Comment by ludralph** [comment]
Hi @AkashVarmaGadiraju 👋 

Thank you for raising this question.

We do have a confirmed issue that MongoDB queries in Prisma are much slower compared to Mongoose. 

- https://github.com/prisma/prisma/issues/16916

It would be helpful if you could document your use case on that issue so that our engineers can take a look.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by AkashVarmaGadiraju** [reply]
@ludralph, I appreciate your assistance, but the provided link did not contain a satisfactory resolution to my issue. If possible, I would like to work together to resolve this matter. I am willing to provide any necessary information or assistance. Thank you.

**Reply by driptanil** [reply]
Hey, did you find any solution ?
I was facing the same issue

**Reply by ludralph** [reply]
Hi @AkashVarmaGadiraju and @driptanil 

We released improvements to MongoDB since Prisma version 5.14. You can find the relevant discussion [here](https://github.com/prisma/prisma/discussions/23888#discussioncomment-9457086).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by danil-iglu** [comment]
@AkashVarmaGadiraju you can use [https://www.npmjs.com/package/prisma-sql](https://www.npmjs.com/package/prisma-sql), it can help with Postgres and SQLite reads
